portal/flatpak-portal.c

```diff
@@ -700,6 +784,1125 @@ authorize_method_handler (GDBusInterfaceSkeleton *interface,
+}
+
+static GFile *
+update_monitor_get_installation_path (PortalFlatpakUpdateMonitor *monitor)
+{
+  UpdateMonitorData *m = update_monitor_get_data (monitor);
+  g_autoptr(GFile) app_path = NULL;
+
+  app_path = g_file_new_for_path (m->app_path);
+
+  /* The app path is always 6 level deep inside the installation dir,
+   * like $dir/app/org.the.app/x86_64/stable/$commit/files, so we find
+   * the installation by just going up 6 parents. */
+  return g_file_resolve_relative_path (app_path, "../../../../../..");
+}
+
+static void
+check_for_updates (PortalFlatpakUpdateMonitor *monitor)
+{
+  UpdateMonitorData *m = update_monitor_get_data (monitor);
+  g_autoptr(GFile) installation_path = NULL;
+  g_autoptr(FlatpakInstallation) installation = NULL;
+  g_autoptr(FlatpakInstalledRef) installed_ref = NULL;
+  g_autoptr(FlatpakRemoteRef) remote_ref = NULL;
+  const char *origin = NULL;
+  const char *local_commit = NULL;
+  const char *remote_commit;
+  g_autoptr(GError) error = NULL;
+
+  installation_path = update_monitor_get_installation_path (monitor);
+
+  g_debug ("Checking for updates for %s/%s/%s in %s", m->name, m->arch, m->branch, flatpak_file_get_path_cached (installation_path));
+
+  installation = lookup_installation_for_path (installation_path, &error);
+  if (installation == NULL)
+    {
+      g_debug ("Unable to find installation for path %s: %s", flatpak_file_get_path_cached (installation_path), error->message);
+      return;
+    }
+
+  installed_ref = flatpak_installation_get_installed_ref (installation,
+                                                          FLATPAK_REF_KIND_APP,
+                                                          m->name, m->arch, m->branch,
+                                                          m->cancellable, &error);
+  if (installed_ref == NULL)
+    {
+      g_debug ("getting installed ref failed: %s", error->message);
+      return; /* Never report updates for uninstalled refs */
+    }
+
+  local_commit = flatpak_ref_get_commit (FLATPAK_REF (installed_ref));
+
+  origin = flatpak_installed_ref_get_origin (installed_ref);
+
+  remote_ref = flatpak_installation_fetch_remote_ref_sync (installation, origin,
+                                                           FLATPAK_REF_KIND_APP,
+                                                           m->name, m->arch, m->branch,
+                                                           m->cancellable, &error);
+  if (remote_ref == NULL)
+    {
+      /* Probably some network issue.
+       * Fall back to the local_commit to at least be able to pick up already installed updates.
+       */
+      g_debug ("getting remote ref failed: %s", error->message);
+      g_clear_error (&error);
+      remote_commit = local_commit;
+    }
+  else
+    {
+      remote_commit = flatpak_ref_get_commit (FLATPAK_REF (remote_ref));
+      if (remote_commit == NULL)
+        {
+          /* This can happen if we're offline and there is an update from an usb drive.
```

Comment:
Contrary to what I said earlier, `remote_commit` should never be NULL. While working on #3114 I discovered that `flatpak_installation_fetch_remote_ref_sync()` actually just errored out if you're offline, prior to that PR. And after that PR, it gets the commit from offline sources. We could leave the `if (remote_commit == NULL)` check just in case but perhaps update the comment? Sorry for the confusion.